packages/middleware-bucket-endpoint/src/bucketEndpointMiddleware.ts

```diff
@@ -0,0 +1,70 @@
+import { bucketHostname } from "./bucketHostname";
+import { BucketEndpointResolvedConfig } from "./configurations";
+import {
+  BuildHandler,
+  BuildHandlerArguments,
+  BuildHandlerOptions,
+  BuildHandlerOutput,
+  BuildMiddleware,
+  MetadataBearer,
+  Pluggable
+} from "@aws-sdk/types";
+import { HttpRequest } from "@aws-sdk/protocol-http";
+
+export function bucketEndpointMiddleware(
+  options: BucketEndpointResolvedConfig
+): BuildMiddleware<any, any> {
+  return <Output extends MetadataBearer>(
+    next: BuildHandler<any, Output>
+  ): BuildHandler<any, Output> => async (
+    args: BuildHandlerArguments<any>
+  ): Promise<BuildHandlerOutput<Output>> => {
+    const { Bucket: bucketName } = args.input;
+    let replaceBucketInPath =
+      options.preformedBucketEndpoint || options.bucketEndpoint;
```

Comment:
I don't get the difference of `preformedBucketEndpoint` and `bucketEndpoint`, we might need to document them

Reply:
I removed ``preformedBucketEndpoint``. only ``bucketEndpoint`` is needed now that we combined the input and config options.